src/peft/utils/other.py

```diff
@@ -418,19 +538,36 @@ def _freeze_adapter(model, adapter_name):
             p.requires_grad = False
 
 
-def _set_trainable(model, adapter_name, modules_to_save):
+def _set_trainable(model, adapter_name, module_names, mode="retrain", **wrapper_kwargs):
+    """Wraps modules that are supposed to be re-trained either normally, i.e. marking them to require gradients and
+    saving them alongside other modules, or with certain methods that go alongside PEFT methods, such as retraining
+    specific token indices using sparse matrices.
+
+    Note that you need to validate beforehand if there are layers shared between modes, e.g. if the 'embedding' layer
+    is configured for both mode `retrain` and `new_tokens` which would result in conflicts down the line.
+    """
+    if mode == "retrain":
```

Comment:
I wonder: instead of the mode argument, can we pass the `target_cls` directly? If setting `ModulesToSaveWrapper` as a default value results in circular imports, it could be set to `None` and then default to `ModulesToSaveWrapper` here.